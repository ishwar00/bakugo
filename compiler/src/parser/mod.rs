use miette::NamedSource;
use pest::{
    error::Error,
    iterators::{Pair, Pairs},
    Parser,
};

use crate::ast::{BakugoParsingErrorDisplay, Node, SourceFile};

#[derive(pest_derive::Parser)]
#[grammar = "bakugo.pest"]
pub struct BakugoParser;

pub fn parse_string(s: &str) -> Result<Pairs<'_, Rule>, Box<Error<Rule>>> {
    match BakugoParser::parse(Rule::SourceFile, s) {
        Ok(ast) => Ok(ast),
        Err(err) => Err(Box::new(err)),
    }
}

pub fn construct_ast(
    source: NamedSource,
    pair: Pair<'_, Rule>,
) -> Result<SourceFile<'_>, BakugoParsingErrorDisplay> {
    match SourceFile::parse(pair) {
        Ok(s) => Ok(s),
        Err(bakugo_err) => Err(BakugoParsingErrorDisplay::from_error(bakugo_err, source)),
    }
}
