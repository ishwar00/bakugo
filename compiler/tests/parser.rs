use std::fs;

use bakugo::parser::parse_string;
use insta::{assert_snapshot, assert_yaml_snapshot, glob};

#[test]
fn test_parser() {
    glob!("examples/*.bakugo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let parsed = parse_string(&input);
        match parsed {
            Ok(parsed) => assert_yaml_snapshot!(parsed),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    });
}
