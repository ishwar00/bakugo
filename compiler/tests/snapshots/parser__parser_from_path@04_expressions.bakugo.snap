---
source: tests/parser.rs
expression: parsed
input_file: tests/examples/04_expressions.bakugo
---
pos:
  - 0
  - 159
pairs:
  - pos:
      - 0
      - 159
    rule: SourceFile
    inner:
      pos:
        - 0
        - 159
      pairs:
        - pos:
            - 0
            - 157
          rule: TopLevelDecl
          inner:
            pos:
              - 0
              - 157
            pairs:
              - pos:
                  - 0
                  - 157
                rule: FunctionDecl
                inner:
                  pos:
                    - 5
                    - 155
                  pairs:
                    - pos:
                        - 5
                        - 9
                      rule: Ident
                      inner: expr
                    - pos:
                        - 9
                        - 12
                      rule: Signature
                      inner:
                        pos:
                          - 9
                          - 11
                        pairs:
                          - pos:
                              - 9
                              - 11
                            rule: Parameters
                            inner: ()
                    - pos:
                        - 15
                        - 155
                      rule: StatementList
                      inner:
                        pos:
                          - 15
                          - 155
                        pairs:
                          - pos:
                              - 15
                              - 25
                            rule: ExpressionStmt
                            inner:
                              pos:
                                - 15
                                - 25
                              pairs:
                                - pos:
                                    - 15
                                    - 16
                                  rule: UnaryMinus
                                  inner: "-"
                                - pos:
                                    - 16
                                    - 17
                                  rule: IntLit
                                  inner: "4"
                                - pos:
                                    - 18
                                    - 19
                                  rule: Plus
                                  inner: +
                                - pos:
                                    - 20
                                    - 21
                                  rule: IntLit
                                  inner: "2"
                                - pos:
                                    - 22
                                    - 23
                                  rule: Star
                                  inner: "*"
                                - pos:
                                    - 24
                                    - 25
                                  rule: IntLit
                                  inner: "3"
                          - pos:
                              - 25
                              - 26
                            rule: Semicolon
                            inner: ;
                          - pos:
                              - 28
                              - 37
                            rule: ExpressionStmt
                            inner:
                              pos:
                                - 28
                                - 37
                              pairs:
                                - pos:
                                    - 28
                                    - 29
                                  rule: IntLit
                                  inner: "4"
                                - pos:
                                    - 30
                                    - 31
                                  rule: And
                                  inner: "&"
                                - pos:
                                    - 32
                                    - 33
                                  rule: IntLit
                                  inner: "3"
                                - pos:
                                    - 34
                                    - 35
                                  rule: Pipe
                                  inner: "|"
                                - pos:
                                    - 36
                                    - 37
                                  rule: IntLit
                                  inner: "1"
                          - pos:
                              - 37
                              - 38
                            rule: Semicolon
                            inner: ;
                          - pos:
                              - 40
                              - 41
                            rule: ExpressionStmt
                            inner:
                              pos:
                                - 40
                                - 41
                              pairs:
                                - pos:
                                    - 40
                                    - 41
                                  rule: IntLit
                                  inner: "4"
                          - pos:
                              - 41
                              - 42
                            rule: Semicolon
                            inner: ;
                          - pos:
                              - 44
                              - 64
                            rule: ExpressionStmt
                            inner:
                              pos:
                                - 44
                                - 64
                              pairs:
                                - pos:
                                    - 44
                                    - 45
                                  rule: IntLit
                                  inner: "4"
                                - pos:
                                    - 46
                                    - 47
                                  rule: Plus
                                  inner: +
                                - pos:
                                    - 48
                                    - 61
                                  rule: FunctionCall
                                  inner:
                                    pos:
                                      - 48
                                      - 60
                                    pairs:
                                      - pos:
                                          - 48
                                          - 54
                                        rule: Ident
                                        inner: bakugo
                                      - pos:
                                          - 54
                                          - 60
                                        rule: Arguments
                                        inner:
                                          pos:
                                            - 55
                                            - 59
                                          pairs:
                                            - pos:
                                                - 55
                                                - 59
                                              rule: ExpressionList
                                              inner:
                                                pos:
                                                  - 55
                                                  - 59
                                                pairs:
                                                  - pos:
                                                      - 55
                                                      - 56
                                                    rule: IntLit
                                                    inner: "3"
                                                  - pos:
                                                      - 56
                                                      - 57
                                                    rule: Comma
                                                    inner: ","
                                                  - pos:
                                                      - 58
                                                      - 59
                                                    rule: IntLit
                                                    inner: "4"
                                - pos:
                                    - 61
                                    - 62
                                  rule: Star
                                  inner: "*"
                                - pos:
                                    - 63
                                    - 64
                                  rule: IntLit
                                  inner: "4"
                          - pos:
                              - 64
                              - 65
                            rule: Semicolon
                            inner: ;
                          - pos:
                              - 67
                              - 98
                            rule: ExpressionStmt
                            inner:
                              pos:
                                - 67
                                - 98
                              pairs:
                                - pos:
                                    - 67
                                    - 68
                                  rule: IntLit
                                  inner: "2"
                                - pos:
                                    - 69
                                    - 70
                                  rule: Plus
                                  inner: +
                                - pos:
                                    - 71
                                    - 98
                                  rule: FunctionCall
                                  inner:
                                    pos:
                                      - 71
                                      - 98
                                    pairs:
                                      - pos:
                                          - 71
                                          - 77
                                        rule: Ident
                                        inner: bakugo
                                      - pos:
                                          - 77
                                          - 95
                                        rule: Arguments
                                        inner:
                                          pos:
                                            - 78
                                            - 94
                                          pairs:
                                            - pos:
                                                - 78
                                                - 94
                                              rule: ExpressionList
                                              inner:
                                                pos:
                                                  - 78
                                                  - 94
                                                pairs:
                                                  - pos:
                                                      - 78
                                                      - 79
                                                    rule: IntLit
                                                    inner: "0"
                                                  - pos:
                                                      - 80
                                                      - 81
                                                    rule: Star
                                                    inner: "*"
                                                  - pos:
                                                      - 82
                                                      - 83
                                                    rule: IntLit
                                                    inner: "2"
                                                  - pos:
                                                      - 83
                                                      - 84
                                                    rule: Comma
                                                    inner: ","
                                                  - pos:
                                                      - 85
                                                      - 86
                                                    rule: IntLit
                                                    inner: "0"
                                                  - pos:
                                                      - 86
                                                      - 87
                                                    rule: Comma
                                                    inner: ","
                                                  - pos:
                                                      - 88
                                                      - 89
                                                    rule: UnaryMinus
                                                    inner: "-"
                                                  - pos:
                                                      - 89
                                                      - 90
                                                    rule: IntLit
                                                    inner: "3"
                                                  - pos:
                                                      - 91
                                                      - 92
                                                    rule: Star
                                                    inner: "*"
                                                  - pos:
                                                      - 93
                                                      - 94
                                                    rule: IntLit
                                                    inner: "4"
                                      - pos:
                                          - 95
                                          - 98
                                        rule: Arguments
                                        inner:
                                          pos:
                                            - 96
                                            - 97
                                          pairs:
                                            - pos:
                                                - 96
                                                - 97
                                              rule: ExpressionList
                                              inner:
                                                pos:
                                                  - 96
                                                  - 97
                                                pairs:
                                                  - pos:
                                                      - 96
                                                      - 97
                                                    rule: IntLit
                                                    inner: "3"
                          - pos:
                              - 98
                              - 99
                            rule: Semicolon
                            inner: ;
                          - pos:
                              - 101
                              - 121
                            rule: ExpressionStmt
                            inner:
                              pos:
                                - 101
                                - 121
                              pairs:
                                - pos:
                                    - 101
                                    - 102
                                  rule: IntLit
                                  inner: "4"
                                - pos:
                                    - 103
                                    - 104
                                  rule: Plus
                                  inner: +
                                - pos:
                                    - 105
                                    - 121
                                  rule: FunctionCall
                                  inner:
                                    pos:
                                      - 105
                                      - 121
                                    pairs:
                                      - pos:
                                          - 105
                                          - 111
                                        rule: Ident
                                        inner: bakugo
                                      - pos:
                                          - 111
                                          - 121
                                        rule: Arguments
                                        inner:
                                          pos:
                                            - 112
                                            - 120
                                          pairs:
                                            - pos:
                                                - 112
                                                - 120
                                              rule: ExpressionList
                                              inner:
                                                pos:
                                                  - 112
                                                  - 120
                                                pairs:
                                                  - pos:
                                                      - 112
                                                      - 113
                                                    rule: IntLit
                                                    inner: "3"
                                                  - pos:
                                                      - 114
                                                      - 115
                                                    rule: Star
                                                    inner: "*"
                                                  - pos:
                                                      - 116
                                                      - 117
                                                    rule: IntLit
                                                    inner: "2"
                                                  - pos:
                                                      - 117
                                                      - 118
                                                    rule: Comma
                                                    inner: ","
                                                  - pos:
                                                      - 119
                                                      - 120
                                                    rule: IntLit
                                                    inner: "3"
                          - pos:
                              - 121
                              - 122
                            rule: Semicolon
                            inner: ;
                          - pos:
                              - 125
                              - 154
                            rule: ExpressionStmt
                            inner:
                              pos:
                                - 125
                                - 154
                              pairs:
                                - pos:
                                    - 125
                                    - 126
                                  rule: IntLit
                                  inner: "3"
                                - pos:
                                    - 127
                                    - 128
                                  rule: Plus
                                  inner: +
                                - pos:
                                    - 129
                                    - 154
                                  rule: FunctionCall
                                  inner:
                                    pos:
                                      - 129
                                      - 154
                                    pairs:
                                      - pos:
                                          - 129
                                          - 135
                                        rule: Ident
                                        inner: bakugo
                                      - pos:
                                          - 135
                                          - 154
                                        rule: Arguments
                                        inner:
                                          pos:
                                            - 136
                                            - 153
                                          pairs:
                                            - pos:
                                                - 136
                                                - 153
                                              rule: ExpressionList
                                              inner:
                                                pos:
                                                  - 137
                                                  - 153
                                                pairs:
                                                  - pos:
                                                      - 137
                                                      - 140
                                                    rule: Ident
                                                    inner: int
                                                  - pos:
                                                      - 141
                                                      - 142
                                                    rule: Comma
                                                    inner: ","
                                                  - pos:
                                                      - 143
                                                      - 145
                                                    rule: Ident
                                                    inner: bb
                                                  - pos:
                                                      - 146
                                                      - 147
                                                    rule: Star
                                                    inner: "*"
                                                  - pos:
                                                      - 148
                                                      - 149
                                                    rule: IntLit
                                                    inner: "3"
                                                  - pos:
                                                      - 149
                                                      - 150
                                                    rule: Comma
                                                    inner: ","
                                                  - pos:
                                                      - 151
                                                      - 153
                                                    rule: Ident
                                                    inner: cc
                          - pos:
                              - 154
                              - 155
                            rule: Semicolon
                            inner: ;
        - pos:
            - 157
            - 158
          rule: Semicolon
          inner: ;
        - pos:
            - 159
            - 159
          rule: EOI
          inner: ""

